optuna/study/_optimize.py

```diff
@@ -219,48 +220,35 @@ def _run_trial(
         state = TrialState.FAIL
         func_err = e
         func_err_fail_exc_info = sys.exc_info()
-    else:
-        # TODO(hvy): Avoid checking the values both here and inside `Study.tell`.
-        values, values_conversion_failure_message = _check_and_convert_to_values(
-            len(study.directions), value_or_values, trial.number
-        )
-        if values_conversion_failure_message is not None:
-            state = TrialState.FAIL
-        else:
-            state = TrialState.COMPLETE
 
     if study._storage.is_heartbeat_enabled():
         assert stop_event is not None
         assert thread is not None
         stop_event.set()
         thread.join()
 
-    # `Study.tell` may raise during trial post-processing.
-    try:
-        study.tell(trial, values=values, state=state)
-    except Exception:
-        raise
-    finally:
-        if state == TrialState.COMPLETE:
-            study._log_completed_trial(trial, cast(List[float], values))
-        elif state == TrialState.PRUNED:
-            _logger.info("Trial {} pruned. {}".format(trial.number, str(func_err)))
-        elif state == TrialState.FAIL:
-            if func_err is not None:
-                _logger.warning(
-                    "Trial {} failed because of the following error: {}".format(
-                        trial.number, repr(func_err)
-                    ),
-                    exc_info=func_err_fail_exc_info,
-                )
-            elif values_conversion_failure_message is not None:
-                _logger.warning(values_conversion_failure_message)
-            else:
-                assert False, "Should not reach."
-        else:
-            assert False, "Should not reach."
+    frozen_trial = study.tell(trial, values=value_or_values, state=state)
+
+    if frozen_trial.state == TrialState.COMPLETE:
+        study._log_completed_trial(trial, frozen_trial.values)
+    elif frozen_trial.state == TrialState.PRUNED:
+        _logger.info("Trial {} pruned. {}".format(trial.number, str(func_err)))
+    elif frozen_trial.state == TrialState.FAIL:
+        if func_err is not None:
+            _logger.warning(
+                "Trial {} failed because of the following error: {}".format(
+                    trial.number, repr(func_err)
+                ),
+                exc_info=func_err_fail_exc_info,
+            )
+    else:
+        assert False, "Should not reach."
 
-    if state == TrialState.FAIL and func_err is not None and not isinstance(func_err, catch):
+    if (
+        frozen_trial.state == TrialState.FAIL
+        and func_err is not None
+        and not isinstance(func_err, catch)
+    ):
         raise func_err
     return trial
```

Comment:
You can simplify the internal logic even further (at `_optimize_sequential`, the caller of `_run_trial`) now with the return value of `Study.tell`.
```suggestion
    return frozen_trial
```

Reply:
Please note that the suggested change here is just a part of my suggestion in the comment.

Reply:
Thank you for the suggestion! It really makes sense to me.
aae3a54 aaf638d ad4e8b5 (ad4e8b5 6f413c6 for multi objective) updates the logic using `FrozenTrial`.

Reply:
Thanks a lot, you can simplify it even further. There is a line in `_optimize_sequential` that obtains a frozen trial from the previously returned Trial. Now that we return a frozen trial, we can skip this line.

```
frozen_trial = copy.deepcopy(study._storage.get_trial(trial._trial_id))
```